src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -393,6 +393,63 @@ bool RegionNode::is_unreachable_from_root(const PhaseGVN* phase) const {
   return true; // The Region node is unreachable - it is dead.
 }
 
+bool RegionNode::is_self_loop(Node* n) {
+  Node_List nstack;
+  VectorSet visited;
+  nstack.push(n);
+  visited.set(n->_idx);
+
+  while(nstack.size()) {
+    n = nstack.pop();
+    for(unsigned i = 1; i < n->req() ; i++) {
+      Node* in = n->in(i);
+      if (in == NULL) {
+        continue;
+      }
+      if (visited.test_set(in->_idx)) {
+        return true;
+      } else {
+        nstack.push(in);
+        visited.set(in->_idx);
+      }
+    }
+  }
+
+  return false;
+}
+
+// If a two input non-loop region has dead input
+// edge[s] degenerate any phi node contained within it.
+bool RegionNode::try_phi_disintegration(PhaseGVN *phase) {
+  if (req() != 3 || isa_Loop() || !in(1) || !in(2) ||
+       (!in(1)->is_top() && !in(2)->is_top())) {
+     return false;
+  }
+
+  PhiNode* phi = has_unique_phi();
+  if (!phi) {
+    return false;
+  }
+
+  Node* rep_node = NULL;
+  PhaseIterGVN *igvn = phase->is_IterGVN();
+  if (in(1)->is_top() && !in(2)->is_top()) {
```

Comment:
The Phi-nodes for loops are always normalized - in(1) will be loop-entry and in(2) is the backedge. So if in(1) is top - in(2) will be a self loop.

Reply:
Yes, loop self loop check is no longer needed, removed associated phi disintegration logic also, there was  a problem with the inputs connection exit_region (convergence region after partially in-lined fast path region and stub call slow path region) which has been fixed, it was maligning the graph shape.